### ISSUE TITLE
[Common Lisp] Implement new Concept Exercise: recursion

### ISSUE THREAD

**Issue by TheLostLambda** [exercism/v3#1714]
This issue describes how to implement the `recursion` concept exercise for the Common Lisp track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Learning objectives

- Be comfortable writing recursive functions in Common Lisp
- Understand tail-recursion and tail-call optimisation (TCO)
- Know that TCO is *not* part of the CL standard, but is commonly implemented
- Be able to write traditional divide-and-conquer functions like quicksort
- Be able to write tail-recursive functions (something like factorial)
- Use local-functions to define tail-recursive helper functions

## Out of scope

- More advanced dynamic programming concepts like memoisation
- Low level details about function calling (beyond TCO)
- The details of different sorting algorithms

## Concepts

- `recursion`

## Prerequisites

- `nested-functions`
- `function-definition`
- `expressions`
- `conditionals`
- `arithemtic`
- `basic-lists`
- `cons`

## Resources to refer to

[Quite Extensive Chapter on Recursion](https://www.cs.cmu.edu/~15110-f12/Touretzky-Common-Lisp-ch8.pdf)

### Hints

- [Recursion from Washington EDU](https://courses.cs.washington.edu/courses/cse341/98wi/CurrentQtr/lectures/lisp3.pdf)

### After

- [CL Implementations With TCO](https://0branch.com/notes/tco-cl.html)

## Representer

Nothing yet

## Analyzer

- Be able to identify tail-recursion vs other forms of recursion

## Implementing

It would be good to have a go at defining something simple like the factorial function in a traditionally recursive way before using a local "go" function to turn it into something tail-recursive.

Many sorting algorithms also lend themselves well to recursion, so implementing something like quicksort or mergesort could be enlightening

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.
